File: app/src/main/java/me/devsaki/hentoid/widget/OnZoneTapListener.java
package me.devsaki.hentoid.widget;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;

import androidx.core.view.GestureDetectorCompat;

import me.devsaki.hentoid.R;

/**
 * Zoned tap listener for the image viewer
 */
public class OnZoneTapListener implements View.OnTouchListener {

    /**
     * This view's dimensions are used to determine which zone a tap belongs to
     */
    private final View view;

    private final GestureDetectorCompat gestureDetector;

    private final int pagerTapZoneWidth;


    private Runnable onLeftZoneTapListener;

    private Runnable onRightZoneTapListener;

    private Runnable onMiddleZoneTapListener;

    private Runnable onLongTapListener;

    public OnZoneTapListener(View view, int tapZoneScale) {
        this.view = view;
        Context context = view.getContext();
        gestureDetector = new GestureDetectorCompat(context, new OnGestureListener());
        pagerTapZoneWidth = context.getResources().getDimensionPixelSize(R.dimen.tap_zone_width) * tapZoneScale;
    }

    public OnZoneTapListener setOnLeftZoneTapListener(Runnable onLeftZoneTapListener) {
        this.onLeftZoneTapListener = onLeftZoneTapListener;
        return this;
    }

    public OnZoneTapListener setOnRightZoneTapListener(Runnable onRightZoneTapListener) {
        this.onRightZoneTapListener = onRightZoneTapListener;
        return this;
    }

    public OnZoneTapListener setOnMiddleZoneTapListener(Runnable onMiddleZoneTapListener) {
        this.onMiddleZoneTapListener = onMiddleZoneTapListener;
        return this;
    }

    public OnZoneTapListener setOnLongTapListener(Runnable onLongTapListener) {
        this.onLongTapListener = onLongTapListener;
        return this;
    }

    public boolean onSingleTapConfirmedAction(MotionEvent e) {
        if (e.getX() < pagerTapZoneWidth && onLeftZoneTapListener != null) {
            onLeftZoneTapListener.run();
        } else if (e.getX() > view.getWidth() - pagerTapZoneWidth && onRightZoneTapListener != null) {
            onRightZoneTapListener.run();
        } else {
            if (onMiddleZoneTapListener != null) onMiddleZoneTapListener.run();
            else return false;
        }
        return true;
    }

    public void onLongPressAction(MotionEvent e) {
        if (onLongTapListener != null) onLongTapListener.run();
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouch(View v, MotionEvent event) {
        return gestureDetector.onTouchEvent(event);
    }

    private final class OnGestureListener extends GestureDetector.SimpleOnGestureListener {
        @Override
        public boolean onSingleTapConfirmed(MotionEvent e) {
            return onSingleTapConfirmedAction(e);
        }

        @Override
        public void onLongPress(MotionEvent e) {
            onLongPressAction(e);
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/widget/OnZoneTapListener.kt
package me.devsaki.hentoid.widget

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import java.lang.ref.WeakReference

/**
 * Zoned tap listener for the reader
 */
class OnZoneTapListener(view: View, private val tapZoneScale: Int) : OnTouchListener {
    /**
     * This view's dimensions are used to determine which zone a tap belongs to
     */
    private val gestureDetector: ViewZoomGestureListener

    private val view: WeakReference<View>

    private var onLeftZoneTapListener: Runnable? = null

    private var onRightZoneTapListener: Runnable? = null

    private var onMiddleZoneTapListener: Runnable? = null

    private var onLongTapListener: Runnable? = null

    init {
        gestureDetector = ViewZoomGestureListener(view.context, GestureListener())
        this.view = WeakReference(view)
    }

    private fun getPagerTapZoneWidth(parentWidth: Int): Int {
        return (parentWidth * 0.18 * tapZoneScale).toInt()
    }

    fun setOnLeftZoneTapListener(onLeftZoneTapListener: Runnable?): OnZoneTapListener {
        this.onLeftZoneTapListener = onLeftZoneTapListener
        return this
    }

    fun setOnRightZoneTapListener(onRightZoneTapListener: Runnable?): OnZoneTapListener {
        this.onRightZoneTapListener = onRightZoneTapListener
        return this
    }

    fun setOnMiddleZoneTapListener(onMiddleZoneTapListener: Runnable?): OnZoneTapListener {
        this.onMiddleZoneTapListener = onMiddleZoneTapListener
        return this
    }

    fun setOnLongTapListener(onLongTapListener: Runnable?): OnZoneTapListener {
        this.onLongTapListener = onLongTapListener
        return this
    }

    fun onSingleTapConfirmedAction(e: MotionEvent): Boolean {
        view.get()?.let {
            if (e.rawX < getPagerTapZoneWidth(it.width) && onLeftZoneTapListener != null) {
                onLeftZoneTapListener?.run()
            } else if (e.rawX > it.width - getPagerTapZoneWidth(it.width) && onRightZoneTapListener != null) {
                onRightZoneTapListener?.run()
            } else {
                if (onMiddleZoneTapListener != null) onMiddleZoneTapListener?.run()
                else return false
            }
        }
        return true
    }

    fun onLongPressAction() {
        onLongTapListener?.run()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        return gestureDetector.onTouchEvent(event)
    }

    inner class GestureListener : ViewZoomGestureListener.Listener() {
        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            return onSingleTapConfirmedAction(e)
        }

        override fun onLongTapConfirmed(ev: MotionEvent) {
            onLongPressAction()
        }
    }
}